File: day6.py
# Decription of script ---------------------------------------------------------
# Day 6 advent of code -- customs questionairres -- what questions were answered 
# and which ones were answered by the whole group
# 
# For this example I went through @ASpittel's solution 
# ------------------------------------------------------------------------------

from collections import Counter
import numpy as np

# solve puzzle ------------------------------------------------------------------

def get_group_count(group):
    group_count = 0
    # count how many times a letter appears in the group
    letter_counts = Counter(group)
    # if the number of times a letter appears is equal to the group size then 
    # all of the group answered yes 
    for letter in letter_counts:
        if letter_counts[letter] == group_size: 
            group_count += 1
    return group_count


with open('day6.txt') as custom_forms:
    # initialise groups and group size 
    group = []
    group_all = []
    group_any = []
    group_size = 0

    for line in custom_forms:
        
        if line != "\n" : # separate groups are on separate lines 
            group_size += 1
            for letter in line.rstrip():
                group.append(letter) 
        else: 
            group_all.append(get_group_count(group)) 
            group_any.append(len(set(group))) # count if anyone has answered a question  (this gets the unique qs)

            group_size = 0 #reset group 
            group = [] # reset group 

    # add the last group 
    group_all.append(len(set(group)))       
    group_any.append(get_group_count(group))

    print("Part 1", sum(group_any))
    print("Part 2", sum(group_all))

File: day6.R
# Decription of script ---------------------------------------------------------
# Day 6 advent of code - customs forma questionaire
# ------------------------------------------------------------------------------

# set up -----------------------------------------------------------------------
library(tidyverse)

# load data --------------------------------------------------------------------
customs_form <-  data.frame(answers = str_sub(read_file(paste0(here::here(), "/day6.txt")), end = -2)) # remove the trailing new line

# solve the puzzle--------------------------------------------------------------
# part 1 - how many unique questions did each group answer year? (a letter means they answered yes)
count_unique <- function(answer){
  length(unique(str_split(answer, "")[[1]]))
}

count_all_answered <- function(answer){
  # the first string in table is \n which represents a new person (this plus 1 is the number of people in a group)
  count_questions <- table(str_split(answer, "")[[1]])
  group_size = ifelse(!any(names(count_questions) == "\n"), 1, count_questions[names(count_questions) == "\n"] + 1)
  sum(count_questions/group_size == 1)
}

customs_form %>%
  separate_rows(answers, sep = "\n\n") %>%
  mutate(answers = str_replace_all(answers, "\n", "")) %>%
  mutate(unique_questions = unlist(purrr::map(answers, count_unique))) %>%
  summarise(sum(unique_questions))

# part 2
# You don't need to identify the questions to which anyone answered "yes"; you need to identify the questions to which everyone answered "yes"!

customs_form %>%
  separate_rows(answers, sep = "\n\n") %>%
  mutate(questions_all_answered = unlist(purrr::map(answers, count_all_answered))) %>%
  summarise(sum(questions_all_answered))
